pkg/dom0-ztools/efi/main.c

```diff
@@ -0,0 +1,697 @@
+{
+  //
+  // The string arrays starts after the SMBIOS_TABLE_TYPE11 structure and
+  // strings are placed one after another with a NULL terminator, so we
+  // start by getting the first one and then move to the next one.
+  //
+  CHAR8 *StringArea = (CHAR8 *)(&Record->StringCount) + sizeof(Record->StringCount);
+  for (UINT8 i = 0; i < Index; i++)
+    StringArea += strlena(StringArea) + 1;
+
+  return StringArea;
+}
+
+EFI_STATUS
+GetSmbiosConfigString(
+    CHAR16 *ConfigName,
+    CHAR16 **ConfigValue)
+{
+  EFI_STATUS Status;
+  EFI_SMBIOS_PROTOCOL *Smbios;
+  SMBIOS_STRUCTURE *Record;
+  SMBIOS_TABLE_TYPE11 *Type11Record;
+  UINT16 Handle;
+  *ConfigValue = NULL;
+
+  Status = uefi_call_wrapper(BS->LocateProtocol,
+                             3,
+                             &gEfiSmbiosProtocolGuid,
+                             NULL,
+                             (VOID **)&Smbios);
+  if (EFI_ERROR(Status))
+    return Status;
+
+  Handle = SMBIOS_HANDLE_PI_RESERVED;
+  while (SmbiosGetNext(Smbios, &Handle, &Record) != EFI_NOT_FOUND)
+  {
+    //
+    // We are only interested in OEM Strings.
+    //
+    if (Record->Type != SMBIOS_TYPE_OEM_STRINGS)
+      continue;
+
+    //
+    // Interpret Record as SMBIOS_TABLE_TYPE11, we can still access the
+    // SMBIOS_STRUCTURE as it is part of the SMBIOS_TABLE_TYPE11.
+    //
+    Type11Record = (SMBIOS_TABLE_TYPE11 *)Record;
+    for (UINT8 i = 0; i < Type11Record->StringCount; i++)
+    {
+      CHAR8 *AsciiOemString = GetSmbiosString(Type11Record, i);
+      CHAR16 *UnicodeOemString = PoolPrint(L"%a", AsciiOemString);
+      if (MetaiMatch(UnicodeOemString, ConfigName))
+      {
+        *ConfigValue = StrSplit(UnicodeOemString, CONFIG_DELIMITER);
+        if (*ConfigValue != NULL)
+          *ConfigValue = PoolPrint(L"%s", *ConfigValue);
+
+        FreePool(UnicodeOemString);
+        return (*ConfigValue != NULL) ? EFI_SUCCESS : EFI_NOT_FOUND;
+      }
+
+      FreePool(UnicodeOemString);
+    }
+  }
+
+  return EFI_NOT_FOUND;
+}
+
+EFI_STATUS
+GetResolutionConfig(
+    UINT32 *HorizontalResolution,
+    UINT32 *VerticalResolution)
+{
+  EFI_STATUS Status;
+  CHAR16 *ConfigValue;
+  CHAR16 *Horizontal;
+  CHAR16 *Vertical;
+
+  Status = GetSmbiosConfigString(EVE_FML_RESOUTION, &ConfigValue);
+  if (EFI_ERROR(Status))
+    return Status;
+
+  Vertical = StrSplit(ConfigValue, L'x');
+  if (Vertical == NULL)
+  {
+    Status = EFI_NOT_FOUND;
+    goto _Return;
+  }
+  *VerticalResolution = Atoi(Vertical);
+
+  ConfigValue[Vertical - ConfigValue - 1] = L'\0';
+  Horizontal = ConfigValue;
+  if (*Horizontal == L'\0')
+  {
+    Status = EFI_NOT_FOUND;
+    goto _Return;
+  }
+  *HorizontalResolution = Atoi(ConfigValue);
+
+_Return:
+  FreePool(ConfigValue);
+  return Status;
+}
+
+EFI_STATUS
+GetBootConfig(
+    BOOLEAN *TryBootOptios)
+{
+  EFI_STATUS Status;
+  CHAR16 *ConfigValue;
+  *TryBootOptios = FALSE;
+
+  Status = GetSmbiosConfigString(EVE_TRY_ALL_BOOT, &ConfigValue);
```

Comment:
@rucoder (2/2) here I convert the smbios OEM string to the a predefined config value and then apply it based on what it is.